nemo/collections/common/tokenizers/canary_tokenizer.py

```diff
@@ -73,7 +73,53 @@ def text_to_ids(self, text, lang_id) -> list[int]:
             return self._tokenize_special_prompt(text)
         lang_id = _map_canary1_to_canary2_lang(lang_id, self.langs)
         if text.endswith(CANARY_EOS):
-            return super().text_to_ids(text[: -len(CANARY_EOS)], lang_id) + [self.eos_id]
+            text_no_eos = text[: -len(CANARY_EOS)]
+            time_pattern = re.compile(r"<\|\d+\|>")
+            time_text = "".join(time_pattern.findall(text_no_eos))
+            has_no_timestamp = not time_text
+            if has_no_timestamp:
+                return super().text_to_ids(text_no_eos, lang_id) + [self.eos_id]
+            else:
+                word_text = time_pattern.sub("", text_no_eos).strip()
+                trans_words = word_text.split()
+
+                # Get timestamp ids
+                time_ids = self._tokenize_special_prompt(time_text)
+
+                # Tokenize text word by wordd
+                word_ids = []
+                result_ids = []
+                time_index = 0
+
+                timestamp_every_n_words = 1  # Add timestmap for every N words
+                word_index = 0
+                # Both start and end time
+                for word in trans_words:
+                    # Insert the first time_id once
+                    if word_index == 0 and time_index < len(time_ids):
+                        result_ids.append(time_ids[time_index])
+                        time_index += 1
+                    # Tokenize the word
+                    word_ids += super().text_to_ids(word, lang_id)
+                    result_ids += super().text_to_ids(word, lang_id)
+                    word_index += 1
+                    # Insert time ids every N words after the first one
+                    if word_index % timestamp_every_n_words == 0 and word_index != 0 and time_index < len(time_ids):
+                        result_ids.append(time_ids[time_index])
+                        time_index += 1
+                        if time_index < len(time_ids):
+                            result_ids.append(time_ids[time_index])
+                            time_index += 1
+                    else:
+                        time_index += 2
+                # Ensure the last time_id is appended at the end
+                if time_index < len(time_ids):
+                    result_ids.append(time_ids[-1])
+                # Make sure the last time_id is appended only once
+                if time_index < len(time_ids) and result_ids[-1] != (time_ids[-1]):
+                    result_ids.append(time_ids[-1])
+                return result_ids + [self.eos_id]
+
         return super().text_to_ids(text, lang_id)
```

Comment:
you missed this branch (when lang_id isn't CANARY_SPECIAL_TOKENIZER and the text doesn't end in EOS) - it also needs to support tokenizing timestamps

Reply:
Thanks for catching that - fixed.